Adding enrollment email exception

stefdworschak opened the pull request:
**Description:** 
Adding logic to catch any exceptions arising from the enrollment script and sending either a success (if no exception was raised) or an error message via email to the platform team to make them aware of the issue.

**ClickUp task:**
[Add error handling to enrollement script to send an email to platform team](https://app.clickup.com/t/5jh4x2)

**Manual Testing:**
Tested by running the `enrollment.py` script from the dev instance with user cidummy as test student; once with a wrong Course of Interest code (STAFF) and once with a correct one.

**Requesits:**
- Add new variable to lms.auth.json called "ZAPIER_ENROLLMENT_EXCEPTION_URL" with the following value: https://hooks.zapier.com/hooks/catch/6213093/ozkkk0o/

Commit:
Implement hotfixes for the enrollment process

stefdworschak committed:
Sending success or exception emails to platform team

stefdworschak committed:
fixing non-keyword arg error

stefdworschak committed:
removing formatted string as this is incompatible with django version

stefdworschak committed:
Fixing indentation level

stefdworschak committed:
Refactoring email content string

stefdworschak committed:
Changing quote format

stefdworschak committed:
changing multiline string

stefdworschak committed:
Adding missing function parameters

stefdworschak committed:
Fixing wrong variable assignment

stefdworschak committed:
Adding some print statements for troubleshooting

stefdworschak committed:
renaming variables

stefdworschak committed:
Removing unneeded variables

yoniLavi commented on lms/djangoapps/student_enrollment/management/commands/enrollment.py:
```diff
@@ -25,6 +26,8 @@
                             'course-v1:CodeInstitute+F101+2017_T1',
                             ]
+FROM_ADDRESS = 'platform@codeinstitute.net'
+TO_ADDRESS = ['platform@codeinstitute.net']
```
If this is intended to allow for multiple addresses, I would go with `TO_ADDRESSES` (in plural). For what it's worth, I just checked and the internal parameter in Django's [send_mail](https://docs.djangoproject.com/en/3.0/topics/email/#send-mail) is named `recipient_list`

stefdworschak replied:
Changed to `RECIPIENT_LIST`

yoniLavi commented on lms/djangoapps/student_enrollment/utils.py:
```diff
@@ -126,3 +129,33 @@ def post_to_zapier(zap_url, data):
+    successful enrollment of students or email that an exception occurred
+    """
+    email_connection = create_email_connection()
+    email_subject = 'Student Enrollment Successful'
```
Functions that take a boolean parameter are often considered a bad idea. In this particular case, `email_type` is a string, which is interpreted as a boolean, which is even stranger. Why not just take `email_subject` directly?

stefdworschak replied:
Changed and removed `email_type` and instead feeding in `email_subject` directly.

yoniLavi commented on lms/djangoapps/student_enrollment/utils.py:
```diff
@@ -126,3 +129,33 @@ def post_to_zapier(zap_url, data):
+        email_subject,
+        content,
+        from_address, to_address,
+        fail_silently=False,
```
`fail_silently=False` raises an exception in case of an error rather than returning zero emails, no?

yoniLavi replied:
If so, it can actually be a nicer approach, as you then don't need to log it - it should just log the error on its own. Also, it really should log an error in this case, rather than the `log.info`

stefdworschak replied:
Changed so that in case of success `log.info` is used and in case of a failure `log.exception`

yoniLavi commented on lms/djangoapps/student_enrollment/management/commands/enrollment.py:
```diff
@@ -45,48 +48,69 @@ def handle(self, *args, **options):
+                                            from_address=FROM_ADDRESS,
+                                            to_address=TO_ADDRESS)
+
+        except Exception as exception:
```
Sorry, but I have a very strong aversion to catching `Exception` , and kinda swallowing it - I would ask to reduce the scope of the `try` and to catch only the actual exception class that is raised by `send_success_or_exception_email` (should be [smtplib.SMTPException](https://docs.python.org/3/library/smtplib.html#smtplib.SMTPException) )

gomyar replied:
This is where Sentry (or other logging provider) can be handy. You can do:

```
try:
    send_mail(...)
except smtplib.SMTPException as se:
    log.exception("There was a problem sending email to %s", to_address)
    raise
```

So the exception gets re-raised (if that's what we want) and the logged exception gets sent to Sentry for future consumption.

stefdworschak replied:
@yoniLavi, not sure I understand this change here. We are trying to catch all errors which prevent the enrollment script from running, but this would only catch errors in sending emails, correct? The idea behind this was that we currently do not have anything in place to tell us whether the enrollment script ran or not and we often times get requests from SC asking us this question. In order to understand if it ran or not we then have to ssh into the server and manually check the log and in the CRM as well. We can change it so that the exception is not caught and we only get an email if the script ran successfully, but I thought having the exception in an email as well would be beneficial and allow us to troubleshoot quicker.

stefdworschak replied:
@yoniLavi and @gomyar, after following up with Nakita there is currently only one real type of exception that we have should expect which is a `ObjectDoesNotExist` exception. Therefore, I changed the enrollment script to catch this exception only. I also added a try...except statement to the method that sends an email in case of an `SMTPException` and am printing the full stacktrace to the log using the `log.exception` statement.

yoniLavi commented on lms/djangoapps/student_enrollment/management/commands/enrollment.py:
```diff
@@ -45,48 +48,69 @@ def handle(self, *args, **options):
+
+                # Set the students access level (i.e. determine whether or
+                # not a student is allowed to access to the LMS.
+                # Deprecated...
```
Do you happen to know what this `Deprecated...` comment is about? It really merits more detail

stefdworschak replied:
No, sorry I have no idea. This must have been from Aaron maybe? I'm not sure if it's really deprecated though, so I created a follow-up task to investigate this: https://app.clickup.com/t/5up74a

yoniLavi replied:
Thanks :)

yoniLavi commented on lms/djangoapps/student_enrollment/management/commands/enrollment.py:
```diff
@@ -45,48 +48,69 @@ def handle(self, *args, **options):
+                    access.allowed_access = True
+                    access.save()
+
+                post_to_zapier(settings.ZAPIER_ENROLLMENT_URL,
```
Let's add a comment here to explain what the Zaps responsibility is - not the specifics of the how, just the what for, so future us know not to recreate that in the LMS

stefdworschak replied:
Added a comment here.

yoniLavi commented on lms/djangoapps/student_enrollment/utils.py:
```diff
@@ -126,3 +129,33 @@ def post_to_zapier(zap_url, data):
+def send_success_or_exception_email(email_type, content, from_address,
+                                    to_address):
+    """
+    Sends an email to the CI platform team either confirming the
```
Given that both the `to_address` and the content are parameters, I would avoid mentioning the specifics in the docstring

stefdworschak replied:
Changed the comment and also changed the parameter to `recipient_list` same as in the enrollment script.

yoniLavi reviewed (changes requested):
I've added some more significant comments than usual, and would suggest reevaluating the approach here before merging
Thanks

stefdworschak committed:
Changing variable name to make it more applicable

Also, changing one of the function parameters to feed in the subject
directly

stefdworschak committed:
Changing logging type based on success or failure

stefdworschak committed:
Adding comments to explain Zapier method

stefdworschak committed:
Merge branch 'master' of https://github.com/Code-Institute-Org/edx-platform into f-enrollment-exception-email

stefdworschak committed:
Moving SMTP error handling to be within the method

Changing the name of one of the parameters to be more
applicable

Changing the log method to exception and sending only the message
via email, but logging the stacktrace as well

stefdworschak committed:
Merge branch 'codeinstitute-ginkgo-stage' of https://github.com/Code-Institute-Org/edx-platform into f-enrollment-exception-email

stefdworschak committed:
Changing back accidentially changed attribute name

yoniLavi commented on lms/djangoapps/student_enrollment/utils.py:
```diff
@@ -126,3 +130,27 @@ def post_to_zapier(zap_url, data):
+def send_success_or_exception_email(email_subject, content, from_address,
+                                    recipient_list):
+    """
+    Sends an email to a list of recipients m either confirming the
```
there's a stray "m" there

yoniLavi commented on lms/djangoapps/student_enrollment/management/commands/enrollment.py:
```diff
@@ -45,48 +52,75 @@ def handle(self, *args, **options):
+                from_address=FROM_ADDRESS,
+                recipient_list=RECIPIENT_LIST)
+
+        except ObjectDoesNotExist as does_not_exist_exception:
```
Thanks @stefdworschak, I much prefer it that we're now catching only a single exception class, but I still feel somewhat uncomfortable about a try... except wrapping such a long piece of code. Where would the `ObjectDoesNotExist` come from, and can we wrap just that?  If it's just the `Program.get`, should we perhaps add some earlier validation to prevent that?

Alternatively, if you want to go back to the approach of catching `Exception`, I quite like Ray's suggestion of re raising it with a `raise` at the end of the `except` clause. But in that case, to still avoid the long indentation of `try...except`, I'd suggest extracting all of the logic above into a separate method named e.g. `attempt_enrollment`, and have `handle` call that and deal with the exception handling.

yoniLavi reviewed (commented):
Hi @stefdworschak, thanks for simplifying this and particularly the `send_success_or_exception_email` at the core of this PR. Now that I can see this more clearly, and looking at the function name, which isn't so relevant anymore with its `or`, I think that this is actually a more useful utility function now, that has nothing to do with enrolment as such.

What this does is change the way we interact with long running tasks in our systems in general. Are we sure that we want to start receiving such emails? And if so, should they be addressed to `platform@codeinstitute.net`, or perhaps we should create a separate email for them?

If we go in this direction, I think we should make a thing out of it. Perhaps naming it something like `send_update_to_admins`, put it in a separate utility module, so it can be easily used across all such command functions and have it use the `FROM_ADDRESS` and `RECIPIENT_LIST` directly, rather than taking them as params?

Two other things to think about:
- As an alternative, given that this would be run by a Student Care adviser (right?), why not have the email go directly to them?
- Do we actually trust the LMS's ability to successfully deliver emails in terms of spam? An alternative would be to perhaps send all such emails from Zapier.

Sorry for being difficult Stefan, but I see this as an entirely new approach and I don't feel comfortable rushing into it.

stefdworschak committed:
Changing the exception logic

stefdworschak committed:
Fixing merge conflict and removing unneeded code

stefdworschak committed:
Removing unneeded code

stefdworschak committed:
Remove unneeded import

stefdworschak committed:
Removing unneeded variables

stefdworschak committed:
Removing unneeded comment

stefdworschak committed:
Fix mistake in comment

stefdworschak committed:
Changed variable to be more descriptive

stefdworschak commented:
@yoniLavi could you please have another look at the updated PR?

I have refactored the code and simplified it based on your comments and the team discussion we had around exception handling between Ray, Nakita and I. You should see the following changes in the logic:

1. Only catch the expected exceptions which occur when the` Course_of_Interest_Code` is not following the agreed schema
2. Log the exception (this can then be used as part of the Sentry setup I believe)
3. Send a request to a Zapier Webhook to send an email to SC with the necessary changes needed for the enrollment to work the next time
4. Continue to the next student in order for the rest of the enrollments not to fail as well
5. Repeat as many times as necessary until the last enrollment

Below you can also find a screenshot of an example exception email. It has a link directly to the student profile on the CRM and gives information on what field SC need to amend to make the enrollment work. We could add some extra details around the schema or could show the actual code used as well, but I'd like to trial this with SC as it is and see if that is actually necessary.
![image](https://user-images.githubusercontent.com/31205982/87781359-7414f900-c828-11ea-9568-71bdee503c58.png)
